server/odc-core/src/main/java/com/oceanbase/odc/core/sql/split/SqlSplitter.java

```diff
@@ -833,4 +847,114 @@ private boolean isIdent(int tokenType) {
+
+        public SqlStatementIterator(InputStream input, Charset charset, Class<? extends Lexer> lexerType,
+                String delimiter) {
+            this.reader = new BufferedReader(new InputStreamReader(input, charset));
+            this.lexerType = lexerType;
+            this.delimiter = delimiter;
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (current == null) {
+                current = parseNext();
+            }
+            return current != null;
+        }
+
+        @Override
+        public OffsetString next() {
+            OffsetString next = current;
+            current = null;
+            if (next == null) {
+                next = parseNext();
+                if (next == null) {
+                    throw new NoSuchElementException("No more available sql.");
+                }
+            }
+            return next;
+        }
+
+        @Override
+        public long iteratedBytes() {
+            return iteratedBytes;
+        }
+
+        private OffsetString parseNext() {
+            try {
+                if (!holder.isEmpty()) {
+                    return holder.poll();
+                }
+                String line;
+                while (holder.isEmpty() && (line = reader.readLine()) != null) {
+                    if (firstLine) {
+                        buffer.append(line);
+                        firstLine = false;
+                    } else {
+                        buffer.append(LINE_SEPARATOR_CHAR).append(line);
+                    }
+                    // SqlSplitter is non-reentrant, so we need to create a new one for each loop
+                    SqlSplitter splitter = createSplitter();
+                    sqls = splitter.split(buffer.toString()).stream().map(OffsetString::getStr)
+                            .collect(Collectors.toList());
+                    while (sqls.size() > SQL_STATEMENT_BUFFER_SIZE) {
+                        String sql = sqls.remove(0);
+                        holder.addLast(new OffsetString(0, sql));
+                        int index = buffer.indexOf(sql.substring(0, sql.length() - 1));
+                        buffer.delete(0, index + sql.length());
+                        clearUselessPrefix();
+                        delimiter = splitter.getDelimiter();
+                    }
+                    iteratedBytes += line.getBytes(Charset.defaultCharset()).length + 1;
+                }
+                if (!holder.isEmpty()) {
+                    return holder.poll();
+                }
+                if (sqls.isEmpty()) {
+                    return null;
+                }
+                return new OffsetString(0, sqls.remove(0));
+            } catch (Exception e) {
+                throw new RuntimeException("Failed to parse input. reason: " + e.getMessage(), e);
+            }
+        }
+
+        private void clearUselessPrefix() {
```

Comment:
we'd better not notice this delimiter

Reply:
But some delimiter has different length, SqlSplitter will all change them to `;`, we must deal with it and remove it from buffer.